File: backend/app/models/media.py
"""
Media Asset Database Model & Enums.
"""

from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Optional

from sqlalchemy import BigInteger, DateTime, Enum as SAEnum, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base


class MediaAssetType(str, Enum):
    """Controlled system for all supported media and document asset types."""
    PROFILE_AVATAR = "profile_avatar"
    COMPANY_LOGO = "company_logo"
    SERVICE_IMAGE = "service_image"
    SERVICE_GALLERY = "service_gallery"
    TECHNICIAN_PORTFOLIO = "technician_portfolio"
    TECHNICIAN_CERTIFICATE = "technician_certificate"
    IDENTITY_DOCUMENT = "identity_document"
    BOOKING_BEFORE = "booking_before"
    BOOKING_AFTER = "booking_after"
    BOOKING_ATTACHMENT = "booking_attachment"
    COMPLAINT_ATTACHMENT = "complaint_attachment"
    REVIEW_IMAGE = "review_image"
    PROPERTY_IMAGE = "property_image"
    JOB_RESUME = "job_resume"
    JOB_DOCUMENT = "job_document"


class MediaAsset(Base):
    """
    Central Media Asset model representing uploaded images and documents in Cloudinary.
    """
    __tablename__ = "media_assets"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    owner_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    owner_type: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    
    asset_type: Mapped[MediaAssetType] = mapped_column(
        SAEnum(MediaAssetType, name="media_asset_type_enum", native_enum=False),
        nullable=False,
        index=True,
    )
    
    cloudinary_asset_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    cloudinary_public_id: Mapped[str] = mapped_column(String(255), unique=True, nullable=False, index=True)
    secure_url: Mapped[str] = mapped_column(String(500), nullable=False)
    resource_type: Mapped[str] = mapped_column(String(30), default="image", nullable=False)
    format: Mapped[str] = mapped_column(String(20), nullable=False, default="")
    width: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    height: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    file_size: Mapped[int] = mapped_column(BigInteger, default=0, nullable=False)
    
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

File: frontend/src/api/media.ts
import api from './axios';

export type MediaAssetType =
  | 'profile_avatar'
  | 'company_logo'
  | 'service_image'
  | 'service_gallery'
  | 'technician_portfolio'
  | 'technician_certificate'
  | 'identity_document'
  | 'booking_before'
  | 'booking_after'
  | 'booking_attachment'
  | 'complaint_attachment'
  | 'review_image'
  | 'property_image'
  | 'job_resume'
  | 'job_document';

export interface MediaAssetResponse {
  id: number;
  owner_id: number;
  owner_type: string;
  asset_type: MediaAssetType;
  cloudinary_asset_id?: string;
  cloudinary_public_id: string;
  secure_url: string;
  thumbnail_url?: string;
  resource_type: string;
  format: string;
  width?: number;
  height?: number;
  file_size: number;
  created_at: string;
  updated_at: string;
}

export interface MediaAssetListResponse {
  total: number;
  items: MediaAssetResponse[];
}

export const mediaApi = {
  uploadMedia: async (data: {
    file: File;
    asset_type: MediaAssetType;
    owner_id?: number;
    owner_type?: string;
  }): Promise<MediaAssetResponse> => {
    const formData = new FormData();
    formData.append('file', data.file);
    formData.append('asset_type', data.asset_type);
    if (data.owner_id !== undefined) formData.append('owner_id', data.owner_id.toString());
    if (data.owner_type !== undefined) formData.append('owner_type', data.owner_type);

    const response = await api.post<MediaAssetResponse>('/media/upload', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return response.data;
  },

  getOwnerMedia: async (
    ownerType: string,
    ownerId: number,
    assetType?: MediaAssetType
  ): Promise<MediaAssetListResponse> => {
    const params = assetType ? { asset_type: assetType } : {};
    const response = await api.get<MediaAssetListResponse>(`/media/owner/${ownerType}/${ownerId}`, {
      params,
    });
    return response.data;
  },

  deleteMedia: async (publicId: string): Promise<any> => {
    const response = await api.delete(`/media/${encodeURIComponent(publicId)}`);
    return response.data;
  },
};
